### PR TITLE
FT: Iam for admin routes

### PR DESCRIPTION
**Pull request by alexandre-merle**
- Modify admin routes to be iam compliant
- Remove all admin routes already on aws cli

Should be merge with https://github.com/scality/Arsenal/pull/105 and https://github.com/scality/Vault/pull/392

### CODE REVIEW

File: tests/unit/v2authTests.js [hunk 1]
```diff
@@ -79,8 +78,20 @@ function handler(req, res) {
         const testCaseIndex = processRequest(
             JSON.parse(req.headers.additionaldata));
         if (typeof testCaseIndex === 'number') {
-            makeResponse(res, responseHeaders[testCaseIndex].code,
-                responseHeaders[testCaseIndex].message);
+            if (testCaseIndex === 0) {
+                makeResponse(res, responseHeaders[testCaseIndex].code,
+                    responseHeaders[testCaseIndex].message);
+            } else {
+                let code = 'Unknow';
```
Comment (jmunoznaranjo) [inline tests/unit/v2authTests.js]: `'Unknown'`?

**Comment by jmunoznaranjo** [comment]
:+1:

**Comment by jmunoznaranjo** [comment]
Removing my plus one, after reviewing https://github.com/scality/Vault/pull/392 I think create-access-key should be kept.

**Comment by jmunoznaranjo** [comment]
Giving my :+1: here after putting up with https://github.com/scality/Vault/pull/392

**Comment by jmunoznaranjo** [comment]
@alexandre-merle feel free to merge if I'm offline

File: bin/vaultclient [hunk 2]
```diff
@@ -98,28 +100,13 @@ program
...
-    .option('--password <PASSWORD>')
     .action(action.bind(null, 'create-account', (client, args) => {
         client.createAccount(
-            args.name, { email: args.email, password: args.password },
+            args.name, { email: args.email },
             handleVaultResponse);
     }));
 
 program
-    .command('list-account-users')
```
Comment (michael-zapata) [inline bin/vaultclient]: This was actually a feature requested by CSEs if I remember correctly.

Comment (jmunoznaranjo) [inline bin/vaultclient]: Good catch!

Comment (michael-zapata) [inline bin/vaultclient]: [There it is](http://docs.aws.amazon.com/IAM/latest/APIReference/API_ListAccessKeys.html)

I think this should only be done when the equivalent functionality is supported in vault.

Comment (jmunoznaranjo) [inline bin/vaultclient]: There are three different listing commands in vaultclient:
-list-accounts: not present in AWS API, neither in aws cli
-list-account-users
-list-access-keys

From those three we should keep list-accounts as it's requested by CSEs and not available in the official tools.

Comment (michael-zapata) [inline bin/vaultclient]: Ah yes, this was the one.

File: lib/IAMClient.js [hunk 3]
```diff
@@ -482,6 +336,34 @@ class IAMClient {
...
+            return parseString(ret, {
+                explicitArray: false,
+            }, (err, result) => {
+                if (err) {
+                    return cb(err);
+                }
+                return cb(null, result);
+            });
+        }
+        try {
```
Comment (rahulreddy) [inline lib/IAMClient.js]: Can you move this try catch into it's own function? Also, you may not want to attach `ret` to `err` object as this will be logged.

File: bin/vaultclient (hunk 2 above)
Comment (alexandre-merle) [inline bin/vaultclient]: I don't remove it :) i have keep the list-accounts and remove account-users and access-keys :)

File: lib/IAMClient.js [hunk 4]
```diff
@@ -166,11 +131,16 @@ class IAMClient {
...
     listAccounts(options, callback) {
+        const data = {
+            Action: 'ListAccounts',
+            Version: '2010-05-08',
+        };
         if (typeof options.marker !== 'undefined') {
             const marker = parseInt(options.marker, 10);
             assert.notStrictEqual(isNaN(marker), true,
                 'Marker must be a number');
             assert(marker >= 0, 'Marker must be >= 0');
```
Comment (LaurenSpiegel) [inline lib/IAMClient.js]: Is there a max?

File: lib/IAMClient.js [hunk 5]
```diff
@@ -449,11 +295,19 @@ class IAMClient {
...
 
-        auth.generateV4Headers(req, JSON.stringify(data),
-                               this.accessKey, this.secretKeyValue);
-
         if (method === 'POST' && typeof data === 'object') {
-            req.write(JSON.stringify(data));
+            auth.generateV4Headers(req, data,
+                    this.accessKey, this.secretKeyValue, 'iam');
+            const paramsToSend = [];
+            Object.keys(data).forEach(key => {
```
Comment (rahulreddy) [inline lib/IAMClient.js]: You can use nodejs' `Query String` module for this.
`querystring.stringify` https://nodejs.org/dist/latest-v4.x/docs/api/querystring.html#querystring_querystring_stringify_obj_sep_eq_options
It even accepts `encodeURIComponent` as it's options.

File: lib/IAMClient.js [hunk 6]
```diff
@@ -482,6 +336,34 @@ class IAMClient {
         req.end();
     }
 
+    getObj(ret, cb) {
```
Comment (LaurenSpiegel) [inline lib/IAMClient.js]: Please add jsdoc for this function.

File: lib/IAMClient.js (hunk 4 above)
Comment (alexandre-merle) [inline lib/IAMClient.js]: Yes in vault we have a max :)

File: tests/unit/httpsTests.js [hunk 7]
```diff
@@ -47,27 +50,26 @@ const testClients = [
...
 }
 
 function handler(req, res) {
     if (req.method === 'POST') {
         extractPost(req, (err, body) => {
             if (err || (body && (body.name !== accountName
-                        || body.emailAddress !== accountOptions.email
-                        || body.saltedPwd !== accountOptions.password))) {
+                        || body.emailAddress !== accountOptions.email))) {
+                let code = 'Unknow';
```
Comment (LaurenSpiegel) [inline tests/unit/httpsTests.js]: Unknown

Comment (rahulreddy) [inline tests/unit/httpsTests.js]: It's hard to follow what's going here, can you add a comment explaining it?
Also, `Unknow` => `Unknown`

File: lib/IAMClient.js (hunk 4 above)
Comment (LaurenSpiegel) [inline lib/IAMClient.js]: So, why not just error here?

File: tests/unit/v2authTests.js [hunk 8]
```diff
@@ -79,8 +78,20 @@ function handler(req, res) {
         const testCaseIndex = processRequest(
             JSON.parse(req.headers.additionaldata));
         if (typeof testCaseIndex === 'number') {
-            makeResponse(res, responseHeaders[testCaseIndex].code,
-                responseHeaders[testCaseIndex].message);
+            if (testCaseIndex === 0) {
+                makeResponse(res, responseHeaders[testCaseIndex].code,
+                    responseHeaders[testCaseIndex].message);
+            } else {
+                let code = 'Unknown';
```
Comment (rahulreddy) [inline tests/unit/v2authTests.js]: This is repeated from the previous file. Can you create a `testUtils.js` and put it in there?

File: lib/IAMClient.js (hunk 4 above)
Comment (alexandre-merle) [inline lib/IAMClient.js]: Oh sorry, not in the marker in fact :) the marker is the pagination :)

**Comment by alexandre-merle** [comment]
Updated :)

**Comment by LaurenSpiegel** [comment]
Once all tests pass including end to end with applicable branches 👍

**Comment by alexandre-merle** [comment]
Merging on behalf of @jmunoznaranjo
